python_modules/dagster/dagster/core/storage/runs/sql_run_storage.py

```diff
@@ -276,54 +276,82 @@ def _runs_query(
             query = db.select(subquery_columns).order_by(subquery.c.rank.asc())
             if bucket_by.bucket_limit:
-                query = query.where(subquery.c.rank <= bucket_by.bucket_limit)
+                return query.where(subquery.c.rank <= bucket_by.bucket_limit)
+            return query
+
+        # bucket by tag - this needs special casing if we have tag-based filters, due to how we
+        # calculate tag filtering in the tags table join.   We need to first apply the tag-based
+        # filtering, and then calculate rank of the filtered runs, then apply the bucket limit /
+        # ordering.
+        check.invariant(isinstance(bucket_by, TagBucket))
+
+        if filters.tags:
+            # first apply the filters to the base query, then use a join to filter, then do the
+            # bucket query
+            filtered_query = db.select([RunsTable.c.run_id]).select_from(
+                RunsTable.join(RunTagsTable, RunsTable.c.run_id == RunTagsTable.c.run_id)
+                if filters.tags
+                else RunsTable
+            )
+            filtered_query = self._add_filters_to_query(filtered_query, filters)
+            filtered_query = filtered_query.alias("filtered_query")
+            base_table = RunsTable.join(
+                RunTagsTable, RunsTable.c.run_id == RunTagsTable.c.run_id
+            ).join(filtered_query, RunsTable.c.run_id == filtered_query.c.run_id)
         else:
-            if filters.tags:
-                base_query = db.select(query_columns).select_from(
-                    RunsTable.join(RunTagsTable, RunsTable.c.run_id == RunTagsTable.c.run_id)
-                )
-            else:
-                base_query = db.select(query_columns).select_from(RunsTable)
+            base_table = RunsTable.join(RunTagsTable, RunsTable.c.run_id == RunTagsTable.c.run_id)
+
+        subquery_columns = [getattr(RunsTable.c, column) for column in columns] + [
+            self._bucket_rank_column(bucket_by, order_by, ascending)
+        ]
+        subquery = (
+            db.select(subquery_columns)
+            .select_from(base_table)
+            .where(RunTagsTable.c.key == bucket_by.tag_key)
+            .where(RunTagsTable.c.value.in_(bucket_by.tag_values))
+        )
+        if not filters.tags:
+            # if there are tag filters, they were already applied to the base query
+            subquery = self._add_filters_to_query(subquery, filters)
+        subquery = subquery.alias("subquery")
 
-            query = self._add_filters_to_query(base_query, filters)
-            query = self._add_cursor_limit_to_query(query, cursor, limit, order_by, ascending)
+        query_columns = [getattr(subquery.c, column) for column in columns]
+        query = db.select(query_columns).order_by(subquery.c.rank.asc())
+        if bucket_by.bucket_limit:
+            query = query.where(subquery.c.rank <= bucket_by.bucket_limit)
 
         return query
```

Comment:
<!-- GRAPHITE_METADATA {"trueLocation":{"startLine":259,"endLine":356}} -->
<i class='graphite__hidden'>[Re: lines 259 to 356]<br/><br/></i>

worth staring at this for a bit and considering how to clean it up for readability - becoming quite a monster